test/integration/driver.go

```diff
@@ -243,53 +249,65 @@ func dumpProxyLogs(name string) {
 }
 
 // makeRequests executes requests in pods and collects request ids per pod to check against access logs
-func makeRequests(pods map[string]string) map[string][]string {
+func makeRequests(pods map[string]string) (map[string][]string, error) {
+	var mu sync.Mutex
 	out := make(map[string][]string)
 	for app := range pods {
 		out[app] = make([]string, 0)
 	}
 
+	g, ctx := errgroup.WithContext(context.Background())
 	testPods := []string{"a", "b", "t"}
 	for _, src := range testPods {
 		for _, dst := range testPods {
 			for _, port := range []string{"", ":80", ":8080"} {
 				for _, domain := range []string{"", "." + namespace} {
-					for n := 0; ; n++ {
+					// capture local copy of loop variables
+					src, dst, port, domain := src, dst, port, domain
+					g.Go(func() error {
 						url := fmt.Sprintf("http://%s%s%s/%s", dst, domain, port, src)
-						log.Printf("Making a request %s from %s (attempt %d)...\n", url, src, n)
-						request := shell(fmt.Sprintf("kubectl exec %s -n %s -c app client %s",
-							pods[src], namespace, url))
-						log.Println(request)
-						match := regexp.MustCompile("X-Request-Id=(.*)").FindStringSubmatch(request)
-						if len(match) > 1 {
-							id := match[1]
-							log.Printf("id=%s\n", id)
-							out[src] = append(out[src], id)
-							out[dst] = append(out[dst], id)
-							break
-						}
-
-						if src == "t" && dst == "t" {
-							log.Println("Expected no match")
-							break
-						}
-
-						if n > budget {
-							fail(fmt.Sprintf("Failed to inject proxy from %s to %s (url %s)", src, dst, url))
+						for n := 0; n < budget; n++ {
+							log.Printf("Making a request %s from %s (attempt %d)...\n", url, src, n)
+							request := shell(fmt.Sprintf("kubectl exec %s -n %s -c app client %s",
+								pods[src], namespace, url))
+							log.Println(request)
+							match := regexp.MustCompile("X-Request-Id=(.*)").FindStringSubmatch(request)
+							if len(match) > 1 {
+								id := match[1]
+								log.Printf("id=%s\n", id)
+								mu.Lock()
+								out[src] = append(out[src], id)
+								out[dst] = append(out[dst], id)
+								mu.Unlock()
+								return nil
+							}
+
+							// Expected no match
+							if src == "t" && dst == "t" {
+								return nil
+							}
+
+							select {
+							case <-time.After(time.Second):
+								// try again
+							case <-ctx.Done():
```

Comment:
Yup. The errgroup will cancel the group context if any of the goroutines returns with an error and the `select` here exits the goroutine request early when that happens.